### PR TITLE
[TS migration] Migrate 'ReimbursementAccount' to TypeScript

### PR DESCRIPTION
**Pull request by mateuuszzzzz**
### Fixed Issues

$ https://github.com/Expensify/App/issues/24928
PROPOSAL:

### Tests

- [ ] Go to settings of a workspace that you own (or have admin permissions).
- [ ] Go to Reimbursements section and press `Connect bank account`.
- [ ] Test both options for adding account (plaid and manual).
- [ ] Verify that all steps are present in setup flow.
- [ ] Verify that going backward and forward preserves entered data.
- [ ] Verify if errors appear when incorrect data is entered and if those errors are relevant.
- [ ] Check if canceling (`Start over` button) the flow at any point does not cause problems with future bank account setups.(e.g. if old errors don't appear in future setups and information related to previous setup is removed). 

### Offline tests

- 

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
Not sure if there are any videos to include since the pr modifies action code.
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/60450261/882787d3-3ac0-44ba-91d7-69ec89e2f089

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/60450261/6f925111-df4a-42b7-8c1f-f08f0f9b33e5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/60450261/eefd3534-ba2d-48c0-81c2-9a02df86a254

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/60450261/68daa06b-8bec-4911-ad74-954cee4bca46

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/60450261/25b8c583-4410-47a7-860e-89fb379ac8c7

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/60450261/8114d3a9-33df-418b-836d-04945fa42eba

</details>

### CODE REVIEW

**Comment by mateuuszzzzz** [comment]
I noticed this typescript error:

`Error: src/pages/ReimbursementAccount/VerifyIdentity/VerifyIdentity.tsx(45,59): error TS2353: Object literal may only specify known properties, and 'isOnfidoSetupComplete' does not exist in type 'Partial<ReimbursementAccountForm>'`

Looks like we missed `isOnfidoSetupComplete`. I wonder which place would be valid for it since `ReimbursementAccountForm` is composition of a few types 🤔

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 1]
```diff
@@ -46,7 +48,11 @@ function getBankAccountList() {
 }
 
 function hasCreditBankAccount() {
-    return _.some(bankAccountList, (bankAccountJSON) => {
+    if (!bankAccountList) {
+        return false;
+    }
+
+    Object.entries(bankAccountList).some(([, bankAccountJSON]) => {
```
Comment (blazejkustra) [inline src/libs/actions/ReimbursementAccount/store.ts:55]: Why not Object.values here?

**Review by blazejkustra — APPROVED** [review]
One minor comments, otherwise LGTM!

File: src/libs/actions/ReimbursementAccount/store.ts (hunk 1 above)
Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/store.ts]: Ah yeah, honestly I have no idea 😅 But in fact, values seem a better choice here.

File: src/libs/actions/ReimbursementAccount/index.ts [hunk 2]
```diff
@@ -11,26 +13,20 @@ export {setBankAccountFormValidationErrors, setPersonalBankAccountFormValidation
  * - `null` if we want to go back to the view where the user selects between connecting via Plaid or connecting manually
  * - CONST.BANK_ACCOUNT.SETUP_TYPE.MANUAL to ask them to enter their accountNumber and routingNumber
  * - CONST.BANK_ACCOUNT.SETUP_TYPE.PLAID to ask them to login to their bank via Plaid
- *
- * @param {String | null} subStep
- * @returns {Promise<void>}
  */
-function setBankAccountSubStep(subStep) {
+function setBankAccountSubStep(subStep: BankAccountSubStep | null) {
```
Comment (VickyStash) [inline src/libs/actions/ReimbursementAccount/index.ts:17]: Let's add return type

File: src/libs/actions/ReimbursementAccount/navigation.ts [hunk 3]
```diff
@@ -2,23 +2,22 @@ import Onyx from 'react-native-onyx';
...
     Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {achData: {currentStep: stepID}});
 }
 
 /**
  * Navigate to the correct bank account route based on the bank account state and type
  *
- * @param {string} policyID - The policy ID associated with the bank account.
- * @param {string} [backTo=''] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
+ * @param policyID - The policy ID associated with the bank account.
+ * @param [backTo=''] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
```
Comment (VickyStash) [inline src/libs/actions/ReimbursementAccount/navigation.ts:18]:
```suggestion
 * @param [backTo] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
```

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
-import _ from 'underscore';
 import BankAccount from '@libs/models/BankAccount';
 import ONYXKEYS from '@src/ONYXKEYS';
+import type * as OnyxTypes from '@src/types/onyx';
+import type {ACHData} from '@src/types/onyx/ReimbursementAccount';
+import type {EmptyObject} from '@src/types/utils/EmptyObject';
 
 /** Reimbursement account actively being set up */
-let reimbursementAccountInSetup = {};
+let reimbursementAccountInSetup: ACHData | EmptyObject = {};
```
Comment (VickyStash) [inline src/libs/actions/ReimbursementAccount/store.ts:10]: Maybe we can do the same you did with `credentials` in this file, wdyt? But also I haven't found this values are really being used. It's used only in a function which is used nowhere...
```suggestion
let reimbursementAccountInSetup: OnyxEntry<ACHData> = null;
```

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 5]
```diff
@@ -1,39 +1,41 @@
...
 Onyx.connect({
     key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
     callback: (val) => {
-        reimbursementAccountInSetup = lodashGet(val, 'achData', {});
+        reimbursementAccountInSetup = val?.achData ?? {};
```
Comment (VickyStash) [inline src/libs/actions/ReimbursementAccount/store.ts:14]:
```suggestion
        reimbursementAccountInSetup = val?.achData ?? null;
```

File: src/libs/actions/ReimbursementAccount/index.ts [hunk 6]
```diff
@@ -11,26 +13,20 @@ export {setBankAccountFormValidationErrors, setPersonalBankAccountFormValidation
...
     return Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {achData: {subStep}});
 }
 
 function hideBankAccountErrors() {
     Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {error: '', errors: null});
 }
 
-function setWorkspaceIDForReimbursementAccount(workspaceID) {
+function setWorkspaceIDForReimbursementAccount(workspaceID: string | null) {
```
Comment (kubabutkiewicz) [inline src/libs/actions/ReimbursementAccount/index.ts:25]: I search for a usage for this function and it seems its not used anywhere, maybe we can delete it?

Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/index.ts:25]: @kubabutkiewicz  there are even more functions that are not used anywhere or are only re-exported in other modules but again not used. I'm not sure if this is a mistake and they should be deleted or temporarily those functions are not in use. cc: @VickyStash 

Here's the full list of what I found:

Not used anywhere:

`getReimbursementAccountInSetup`
`getReimbursementAccountWorkspaceID`
`getCredentials`

Exported in `ReimbursementAccount` (`index.ts`) but not used anywhere:

`setPersonalBankAccountFormValidationErrorFields`
`showBankAccountFormValidationError`
`deleteFromBankAccountList`

Exported in `BankAccount.ts` and `ReimbursementAccount` (`index.ts`) but not used anywhere:

`setBankAccountFormValidationErrors`
`setWorkspaceIDForReimbursementAccount`

Comment (kubabutkiewicz) [reply src/libs/actions/ReimbursementAccount/index.ts:25]: @fabioh8010 do you think its safe to remove them?

File: src/libs/actions/ReimbursementAccount/store.ts (hunk 4 above)
Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/store.ts:10]: Hmm but now I think about it and I'm not sure if we use `OnyxEntry` for keys of `Onyx` types.
I mean that `ACHData` is not `OnyxEntry` itself but it is a key of `ReimbursementAccount` 🤔

Comment (VickyStash) [reply src/libs/actions/ReimbursementAccount/store.ts:10]: I'm okay to leave it as is then 👌

File: src/libs/actions/ReimbursementAccount/index.ts (hunk 6 above)
Comment (fabioh8010) [reply src/libs/actions/ReimbursementAccount/index.ts:25]: I think we could check if he had recent activity on these functions, because sometimes we have code that looks unused but it was created with the purpose to be used later (during a feature development for example).

Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/index.ts:25]: Well, those functions seems to be pretty old:

`getReimbursementAccountInSetup` (~2 years) https://github.com/Expensify/App/pull/6230
`getReimbursementAccountWorkspaceID` (~2 years) https://github.com/Expensify/App/pull/6230
`getCredentials` (~2 years) https://github.com/Expensify/App/pull/6230
`setPersonalBankAccountFormValidationErrorFields` (~20 months) https://github.com/Expensify/App/pull/9423
`showBankAccountFormValidationError` (~2 years) https://github.com/Expensify/App/pull/6230
`deleteFromBankAccountList` (~2 years) https://github.com/Expensify/App/pull/7557
`setWorkspaceIDForReimbursementAccount` (~2 years) https://github.com/Expensify/App/pull/6230

Comment (fabioh8010) [reply src/libs/actions/ReimbursementAccount/index.ts:25]: Thanks for the info @mateuuszzzzz ! Well, I suppose we can remove them, let's do it and see how it goes during C+/Exfy reviews.

Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/index.ts]: Cool, I removed them

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+let bankAccountList: OnyxEntry<OnyxTypes.BankAccountList> = null;
+Onyx.connect({
+    key: ONYXKEYS.BANK_ACCOUNT_LIST,
+    callback: (val) => {
+        bankAccountList = val;
+    },
+});
+
+function getBankAccountList() {
```
Comment (kubabutkiewicz) [inline src/libs/actions/ReimbursementAccount/store.ts:15]: add return type

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    return bankAccountList;
+}
+
+function hasCreditBankAccount() {
```
Comment (kubabutkiewicz) [inline src/libs/actions/ReimbursementAccount/store.ts:19]: add return type

File: src/libs/actions/ReimbursementAccount/store.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    if (!bankAccountList) {
+        return false;
+    }
+
+    Object.values(bankAccountList).some((bankAccountJSON) => {
```
Comment (kubabutkiewicz) [inline src/libs/actions/ReimbursementAccount/store.ts:24]: I think you are missing here `return`

Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/store.ts]: Good catch!

**Review by kubabutkiewicz — APPROVED** [review]

**Comment by blazejkustra** [comment]
@mateuuszzzzz I agree with Aimane, let's fill out tests and QA sections

**Review by aimane-chnaif — COMMENTED** [review]
I see that lots of unused functions are removed but was it discussed?

**Review by aimane-chnaif — APPROVED** [review]
Code looks good

**Comment by mountiny** [comment]
@mateuuszzzzz or @blazejkustra can someone merge main please

**Comment by mateuuszzzzz** [comment]
@mountiny @aimane-chnaif done

**Comment by mateuuszzzzz** [comment]
> I see that lots of unused functions are removed but was it discussed?

According to this. We had discussion about deleting them in this PR and decided to do it since those functions are pretty old

**Review by aimane-chnaif — APPROVED** [review]

File: src/libs/actions/ReimbursementAccount/errors.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type {ErrorFields} from '@src/types/onyx/OnyxCommon';
+
+/**
+ * Set the current fields with errors.
+
```
Comment (mountiny) [inline src/libs/actions/ReimbursementAccount/errors.ts:7]:
```suggestion
```

File: src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                // @ts-expect-error: ONYXKEYS.REIMBURSEMENT_ACCOUNT is conflicting with ONYXKEYS.FORMS.REIMBURSEMENT
+                {
+                    onyxMethod: Onyx.METHOD.SET,
+                    key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
+                    value: ReimbursementAccountProps.reimbursementAccountDefaultProps,
+                },
+                {
+                    onyxMethod: Onyx.METHOD.SET,
+                    key: ONYXKEYS.FORMS.REIMBURSEMENT_ACCOUNT_FORM_DRAFT,
+                    value: {
```
Comment (mountiny) [inline src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts:73]: In the js file we just set this to empty object, why this change

File: src/pages/ReimbursementAccount/BankInfo/BankInfo.tsx [hunk 12]
```diff
@@ -118,7 +118,7 @@ function BankInfo({reimbursementAccount, reimbursementAccountDraft, plaidLinkTok
                     [BANK_INFO_STEP_KEYS.ROUTING_NUMBER]: '',
                     [BANK_INFO_STEP_KEYS.ACCOUNT_NUMBER]: '',
                     [BANK_INFO_STEP_KEYS.PLAID_MASK]: '',
-                    [BANK_INFO_STEP_KEYS.IS_SAVINGS]: '',
+                    [BANK_INFO_STEP_KEYS.IS_SAVINGS]: false,
```
Comment (mountiny) [inline src/pages/ReimbursementAccount/BankInfo/BankInfo.tsx:121]: why is this change needed?

File: src/types/form/ReimbursementAccountForm.ts [hunk 13]
```diff
@@ -21,6 +21,7 @@ const INPUT_IDS = {
         CITY: 'requestorAddressCity',
         STATE: 'requestorAddressState',
         ZIP_CODE: 'requestorAddressZipCode',
+        IS_ONFIDO_SETUP_COMPLETE: 'isOnfidoSetupComplete',
```
Comment (mountiny) [inline src/types/form/ReimbursementAccountForm.ts:24]: Why do we need this new input id here?

**Review by mountiny — COMMENTED** [review]
@mateuuszzzzz thanks! Have you tested the full flow in the bank account and added the testing bank account succefully? Then also removed it? 

Seems like there is couple of changes different from the JS so curious why they needed and hence I think we need to be thorough with testing here

File: src/pages/ReimbursementAccount/BankInfo/BankInfo.tsx (hunk 12 above)
Comment (mateuuszzzzz) [reply src/pages/ReimbursementAccount/BankInfo/BankInfo.tsx:121]: Before migration we had incorrect type value here. After migration `typecheck` throws type mismatch so it had to be changed. Since empty string is falsy I assume using `false` boolean value is fine here.

File: src/types/form/ReimbursementAccountForm.ts (hunk 13 above)
Comment (mateuuszzzzz) [reply src/types/form/ReimbursementAccountForm.ts:24]: It turned out we use it here https://github.com/software-mansion-labs/expensify-app-fork/blob/401bb219c74fb615704895554b6876eb25601a11/src/pages/ReimbursementAccount/RequestorOnfidoStep.js#L49 and missed it before

File: src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts (hunk 11 above)
Comment (mateuuszzzzz) [reply src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts:73]: Similarly here `typecheck` throws error, because keys of the type used here are not optional

**Comment by mateuuszzzzz** [comment]
> @mateuuszzzzz thanks! Have you tested the full flow in the bank account and added the testing bank account succefully? Then also removed it?
> 
> Seems like there is couple of changes different from the JS so curious why they needed and hence I think we need to be thorough with testing here

@mountiny  In general, I have a problem with verification via `Onfido` without a social security number, but this is not related to this PR. Similarly for `Plaid` even if the first steps are successful the verification step cannot be passed. Is there any example data that is valid and complete? All examples on open `#expensify-open-source` seem to be outdated or incomplete.

**Comment by kubabutkiewicz** [comment]
image
@mateuuszzzzz you can use those values to pass verification

**Review by mountiny — APPROVED** [review]
Thanks!
